src/hybrid-appliance/azext_hybrid_appliance/custom.py

```diff
@@ -0,0 +1,228 @@
+
+    # The NO_PROXY env variable needs the api server address, which is not available with the user before the cluster is provisioned.
+    # To get around that, we get the server address and services cidr and append the no_proxy variable with those values.
+    if http_proxy or https_proxy or no_proxy:
+        no_proxy = utils.get_proxy_parameters(no_proxy)
+        os.environ["NO_PROXY"] = no_proxy
+    if location:
+        cmd_onboard_arc.extend(["--location", location])
+    if http_proxy:
+        cmd_onboard_arc.extend(["--proxy-http", http_proxy])
+    if https_proxy:
+        cmd_onboard_arc.extend(["--proxy-https", https_proxy])
+    if no_proxy:
+        cmd_onboard_arc.extend(["--proxy-skip-range", no_proxy])
+    if proxy_cert:
+        cmd_onboard_arc.extend(["--proxy-cert", proxy_cert])
+
+    onboarding_result = get_default_cli().invoke(cmd_onboard_arc)
+    if onboarding_result != 0:
+        telemetry.set_exception(exception="Arc onboarding of the k8s cluster failed", fault_type=consts.Arc_Onboarding_Failed, summary="Arc onboarding of the k8s cluster failed")
+        raise CLIInternalError("Onboarding the k8s cluster to Arc failed")
+    else:
+        print("The k8s cluster has been onboarded to Arc successfully")
+
+def upgrade_hybrid_appliance(resource_group_name, name):
+    try:
+        azure_clusterconfig_cm = subprocess.check_output(['microk8s', 'kubectl', 'get', 'cm', 'azure-clusterconfig', '-n', 'azure-arc', '-o', 'json']).decode()
+    except Exception as e:
+        if utils.check_microk8s():
+            # Is there anything else which can be done in this case?
+            telemetry.set_exception(exception=e, fault_type=consts.ConfigMap_not_Found, summary="Configmap 'azure-clusterconfig' not found on the k8s cluster")
+            raise ValidationError("The required configmap 'azure-clusterconfig' was not found on the kubernetes cluster. Exception:" + str(e), "Please delete the appliance and create it again.")
+        else:
+            telemetry.set_exception(exception=e, fault_type=consts.K8s_Cluster_Unreacheable_Or_Not_Running, summary="The k8s cluster is not running as expected or is not reachable")
+            raise ValidationError("The kubernetes cluster is not running as expected or is not reachable. Exception:" + str(e), "Please delete the appliance and create it again.")
+            
+    
+    azure_clusterconfig_cm = json.loads(azure_clusterconfig_cm)
+    try:
+        if azure_clusterconfig_cm["data"]["AZURE_RESOURCE_GROUP"] != resource_group_name or azure_clusterconfig_cm["data"]["AZURE_RESOURCE_NAME"] != name:
+            telemetry.set_exception()
+            raise ValidationError("The parameters passed do not correspond to this appliance", "Please check the resource group name and appliance name.")
+    except KeyError:
+        telemetry.set_exception()
+        raise CLIInternalError("The required entries were not found in the config map.",  "Please delete the appliance and recreate it.")
```

Comment:
this can happen only if user tampers with config map right? then we shouldn't categorise as CLIInternal error. We should rather say validationerror.